### ISSUE TITLE
Rook considering managing the images for MON, OSD, and MGR separately?

### ISSUE THREAD

**Issue by zeaone** [rook/rook#14965]
Is Rook considering managing the images for MON, OSD, and MGR separately? This way, when there is a need to upgrade the MGR, the OSDs do not need to be restarted.

**Comment by microyahoo** [comment]
https://blog.rook.io/rook-v1-13-storage-enhancements-c551f62892bb

> To allow for advanced maintenance and troubleshooting of Ceph daemons, the label “ceph.rook.io/do-not-reconcile” is now respected for all Ceph daemons. If this label is found on a daemon while the operator is reconciling, the daemon will be skipped. Previously, this label was only available to mons and OSDs.

**Comment by travisn** [comment]
Yes, the `do-not-reconcile` label is the recommended way to temporarily skip reconcile. However, for upgrades, it is not generally recommended. If a pod upgrade is skipped, there might be an important update missed. If it's a rook upgrade, there could be a new pod spec setting. If a ceph upgrade, the ceph daemons should all be upgraded without being skipped. But if it's only a rook upgrade and no changes to the OSD spec, the OSDs are not expected to restart.

**Comment by zeaone** [comment]
Currently, our storage cluster is quite large, with approximately 2000 OSDs. If a certain upgrade only requires changes to the MGR, meaning only the MGR needs to be restarted, other components (OSDs, MONs) do not need to be restarted, which would otherwise cause instability in the cluster.

**Comment by travisn** [comment]
Are you seeing OSDs restart on every upgrade? If so, this is a bug. We don't expect the OSDs to restart during upgrade unless necessary. 

Common bugs for the restart is that something changes unintentionally in the pod spec, such as the order of some properties. If you can get the OSD pod spec before the upgrade, then the OSD pod spec after the upgrade, we can diff and see what changed unexpectedly.

If the Ceph version changes, this is always a change that causes an OSD restart. Do you also hope to skip restart in this scenario?

**Comment by travisn** [comment]
During such an update, is there a new Ceph image? For example, you might be updating in the future from v19.2.0 to v19.2.1, and you only want to restart a subset of the ceph daemons? Because I would expect all daemons to be affected by such a ceph version change. Small scoped ceph releases are rare these days. So decoupling the ceph version for each daemon seems risky, if you end up running different versions of ceph for each daemon. 

Or are you making manual changes to ceph, and the ceph version is not changing? In that case, Rook should not anyway restart the ceph daemons, so I assume the ceph version is changing.

**Comment by zeaone** [comment]
A new image will be generated. Assume that currently all my daemons are at version v19.2.1, but I want to modify some functionality on the mgr component (only involving the mgr component, without affecting mon, osd, etc.). To use the new functionality, it involves a version upgrade. If I follow the current upgrade method, I would need to upgrade all unrelated components like mon and osd (which involves restarting the processes). If the images are separate, I can simply upgrade the mgr to use the new functionality.

**Comment by travisn** [comment]
Ok, sounds like you are building custom images of ceph that may have more scoped changes than the upstream releases. Then perhaps it could be implemented in the cephcluster CR with an image for each component:
```
spec:
  cephVersion:
    image: quay.io/ceph/ceph:v18.2.4
    # a specific image to specify for each ceph daemon, in case of advanced scenarios where a subset of daemons are intentionally updated, while others are not
    imageOverrides:
      mon: <ceph-mon-image>
      mgr: <ceph-mgr-image>
      osd: <ceph-osd-image>
      ...
```

**Comment by zeaone** [comment]
yes thanks  iwill try

**Comment by travisn** [comment]
> yes thanks iwill try

Ok, you are going to open a PR for this? We can reopen this issue until it is fixed.
